File: apple/Sources/ReactNativeWatchOS/Modifiers/AlignmentGuide.swift
import SwiftUI

/// SwiftUI `.alignmentGuide(_:computeValue:)`.
///
/// LIMITED: the real API takes an arbitrary `(ViewDimensions) -> CGFloat`
/// closure, which can't cross the JS bridge. This implements a simplified
/// form: a named `guide` plus a numeric `offset`, where `computeValue`
/// returns the guide's default value (`dimension[guide]`) plus the offset.
/// Custom guides and dimension-relative math are not expressible here.
enum RNWAlignmentGuideModifier {
    @MainActor
    static func register(into r: RNWModifierRegistry) {
        r.register("alignmentGuide") { view, params, _ in
            let offset = params.cgFloat("offset") ?? 0
            switch params.string("guide") {
            case "leading":
                return AnyView(view.alignmentGuide(.leading) {
                    $0[.leading] + offset
                })
            case "trailing":
                return AnyView(view.alignmentGuide(.trailing) {
                    $0[.trailing] + offset
                })
            case "centerHorizontal":
                return AnyView(view.alignmentGuide(HorizontalAlignment.center) {
                    $0[HorizontalAlignment.center] + offset
                })
            case "top":
                return AnyView(view.alignmentGuide(.top) {
                    $0[.top] + offset
                })
            case "bottom":
                return AnyView(view.alignmentGuide(.bottom) {
                    $0[.bottom] + offset
                })
            case "centerVertical":
                return AnyView(view.alignmentGuide(VerticalAlignment.center) {
                    $0[VerticalAlignment.center] + offset
                })
            case "firstTextBaseline":
                return AnyView(view.alignmentGuide(.firstTextBaseline) {
                    $0[.firstTextBaseline] + offset
                })
            case "lastTextBaseline":
                return AnyView(view.alignmentGuide(.lastTextBaseline) {
                    $0[.lastTextBaseline] + offset
                })
            default:
                return view
            }
        }
    }
}

File: src/modifiers/alignmentGuide.ts
import { createModifier } from './createModifier';

export type AlignmentGuideName =
  | 'leading'
  | 'trailing'
  | 'top'
  | 'bottom'
  | 'centerHorizontal'
  | 'centerVertical'
  | 'firstTextBaseline'
  | 'lastTextBaseline';

export interface AlignmentGuideParams {
  /// Which alignment guide to override. Horizontal guides
  /// (`leading`/`trailing`/`centerHorizontal`) pair with a parent's
  /// horizontal alignment; vertical guides (`top`/`bottom`/baselines/
  /// `centerVertical`) pair with a vertical one.
  guide: AlignmentGuideName;
  /// Offset added to the guide's default position, in points. The native
  /// side computes `defaultGuideValue + offset`.
  offset?: number;
}

/// SwiftUI `.alignmentGuide(_:computeValue:)`.
///
/// LIMITED: SwiftUI's real `computeValue` is an arbitrary closure over the
/// view's `ViewDimensions`. A closure cannot cross the JS bridge, so this
/// exposes a simplified form — pick a named `guide` and a numeric `offset`,
/// and the native side returns the guide's default value plus that offset.
/// Custom alignment guides and dimension-relative math are not expressible.
export function alignmentGuide(params: AlignmentGuideParams) {
  return createModifier('alignmentGuide', params);
}
